File: 2026-1/sesion-de-clase-07-08/SoftProg/SoftProgPersistencia/src/main/java/pe/edu/pucp/softprog/dao/almacen/ProductoDAOImpl.java
package pe.edu.pucp.softprog.dao.almacen;

import pe.edu.pucp.softprog.db.DBFactoryProvider;
import pe.edu.pucp.softprog.db.DBManager;
import pe.edu.pucp.softprog.modelo.almacen.Producto;
import pe.edu.pucp.softprog.modelo.almacen.UnidadMedida;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class ProductoDAOImpl implements ProductoDAO {

    @Override
    public Integer crear(Producto modelo) {
        DBManager dbManager = DBFactoryProvider.getManager();
        try (Connection conn = dbManager.getConnection();
             PreparedStatement cmd = this.comandoCrear(conn, modelo)){
            if (cmd.executeUpdate() == 0) {
                return null;
            }
            try (ResultSet rs = cmd.getGeneratedKeys()) {
                return rs.next() ? rs.getInt(1) : null;
            }
        }
        catch (SQLException e) {
            System.err.println("Error SQL: " + e.getMessage());
            throw new RuntimeException(e);
        }
        catch (Exception e) {
            System.err.println("Error inesperado: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    @Override
    public boolean actualizar(Producto modelo) {
        DBManager dbManager = DBFactoryProvider.getManager();
        try (Connection conn = dbManager.getConnection();
             PreparedStatement cmd = this.comandoActualizar(conn, modelo)) {
            return cmd.executeUpdate() > 0;
        }
        catch (SQLException e) {
            System.err.println("Error SQL: " + e.getMessage());
            throw new RuntimeException(e);
        }
        catch (Exception e) {
            System.err.println("Error inesperado: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    @Override
    public boolean eliminar(int id) {
        DBManager dbManager = DBFactoryProvider.getManager();
        try (Connection conn = dbManager.getConnection();
             PreparedStatement cmd = this.comandoEliminar(conn, id)) {
            return cmd.executeUpdate() > 0;
        }
        catch (SQLException e) {
            System.err.println("Error SQL: " + e.getMessage());
            throw new RuntimeException(e);
        }
        catch (Exception e) {
            System.err.println("Error inesperado: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    @Override
    public Producto leer(int id) {
        DBManager dbManager = DBFactoryProvider.getManager();
        try (Connection conn = dbManager.getConnection();
             PreparedStatement cmd = this.comandoLeer(conn, id)) {
            ResultSet rs = cmd.executeQuery();

            if (!rs.next()) {
                System.err.println("No se encontro el registro con "
                        + "id: " + id);
                return null;
            }

            return this.mapearModelo(rs);
        }
        catch (SQLException e) {
            System.err.println("Error SQL: " + e.getMessage());
            throw new RuntimeException(e);
        }
        catch (Exception e) {
            System.err.println("Error inesperado: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    @Override
    public java.util.List<Producto> leerTodos() {
        DBManager dbManager = DBFactoryProvider.getManager();
        try (Connection conn = dbManager.getConnection();
             PreparedStatement cmd = this.comandoLeerTodos(conn)) {
            ResultSet rs = cmd.executeQuery();

            List<Producto> modelos = new ArrayList<>();
            while (rs.next()) {
                modelos.add(this.mapearModelo(rs));
            }

            return modelos;
        }
        catch (SQLException e) {
            System.err.println("Error SQL: " + e.getMessage());
            throw new RuntimeException(e);
        }
        catch (Exception e) {
            System.err.println("Error inesperado: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    protected PreparedStatement comandoCrear(Connection conn,
                                             Producto modelo) throws SQLException {
        String sql = """
            INSERT INTO PRODUCTO (
                nombre,
                unidadMedida,
                precio,
                activo
            ) VALUES (?, ?, ?, ?)
            """;

        PreparedStatement cmd = conn.prepareStatement(sql,
                Statement.RETURN_GENERATED_KEYS);
        cmd.setString(1, modelo.getNombre());
        cmd.setString(2,modelo.getUnidadMedida().name());
        cmd.setDouble(3, modelo.getPrecio());
        cmd.setBoolean(4, modelo.isActivo());

        return cmd;
    }

    protected PreparedStatement comandoActualizar(Connection conn,
                                                  Producto modelo) throws SQLException {
        String sql = """
            UPDATE PRODUCTO
            SET nombre = ?,
                unidadMedida = ?,
                precio = ?,
                activo = ?
            WHERE id = ?
            """;

        PreparedStatement cmd = conn.prepareStatement(sql);
        cmd.setString(1, modelo.getNombre());
        cmd.setString(2, modelo.getUnidadMedida().name());
        cmd.setDouble(3, modelo.getPrecio());
        cmd.setBoolean(4, modelo.isActivo());
        cmd.setInt(5, modelo.getId());

        return cmd;
    }

    protected PreparedStatement comandoEliminar(Connection conn,
                                                Integer id) throws SQLException {
        String sql = """
            DELETE FROM PRODUCTO WHERE id = ?
            """;
        PreparedStatement cmd = conn.prepareStatement(sql);
        cmd.setInt(1, id);

        return cmd;
    }

    protected PreparedStatement comandoLeer(Connection conn,
                                            Integer id) throws SQLException {
        String sql = """
            SELECT * FROM PRODUCTO WHERE id = ?
            """;
        PreparedStatement cmd = conn.prepareStatement(sql);
        cmd.setInt(1, id);
        return cmd;
    }

    protected PreparedStatement comandoLeerTodos(Connection conn) throws SQLException {
        String sql = """
            SELECT * FROM PRODUCTO
            """;
        return conn.prepareStatement(sql);
    }

    protected Producto mapearModelo(ResultSet rs) throws SQLException {
        Producto producto = new Producto();
        producto.setId(rs.getInt("id"));
        producto.setNombre(rs.getString("nombre"));
        producto.setUnidadMedida(
                UnidadMedida.valueOf(rs.getString("unidadMedida")));
        producto.setPrecio(rs.getDouble("precio"));
        producto.setActivo(rs.getBoolean("activo"));
        return producto;
    }
}

File: 2026-1/sesion-de-clase-12/con-transacciones-sql/SoftProgPersistencia/Dao/Almacen/ProductoDaoImpl.cs
using System.Data.Common;
using SoftProgModelo.Modelos.Almacen;

namespace SoftProgPersistencia.Dao.Almacen;

public class ProductoDaoImpl : DefaultBaseDao<Producto>, IProductoDao
{
    protected override DbCommand ComandoCrear(DbConnection conn, Producto modelo)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = @"INSERT INTO PRODUCTO (nombre, unidadMedida, precio, activo)
                            VALUES (@p1, @p2, @p3, @p4);";
        CrearParametro(cmd, "@p1", modelo.Nombre);
        CrearParametro(cmd, "@p2", modelo.UnidadMedida.ToString());
        CrearParametro(cmd, "@p3", modelo.Precio);
        CrearParametro(cmd, "@p4", modelo.Activo);
        return cmd;
    }

    protected override DbCommand ComandoActualizar(DbConnection conn, Producto modelo)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = @"UPDATE PRODUCTO SET nombre = @p1, unidadMedida = @p2, precio = @p3, activo = @p4
                            WHERE id = @p5;";
        CrearParametro(cmd, "@p1", modelo.Nombre);
        CrearParametro(cmd, "@p2", modelo.UnidadMedida.ToString());
        CrearParametro(cmd, "@p3", modelo.Precio);
        CrearParametro(cmd, "@p4", modelo.Activo);
        CrearParametro(cmd, "@p5", modelo.Id);
        return cmd;
    }

    protected override DbCommand ComandoEliminar(DbConnection conn, int id)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = "DELETE FROM PRODUCTO WHERE id = @p1;";
        CrearParametro(cmd, "@p1", id);
        return cmd;
    }

    protected override DbCommand ComandoLeer(DbConnection conn, int id)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = "SELECT * FROM PRODUCTO WHERE id = @p1;";
        CrearParametro(cmd, "@p1", id);
        return cmd;
    }

    protected override DbCommand ComandoLeerTodos(DbConnection conn)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = "SELECT * FROM PRODUCTO;";
        return cmd;
    }

    protected override Producto MapearModelo(DbDataReader reader)
    {
        return new Producto
        {
            Id = LeerEntero(reader, "id"),
            Nombre = LeerTexto(reader, "nombre"),
            UnidadMedida = Enum.Parse<UnidadMedida>(LeerTexto(reader, "unidadMedida")),
            Precio = LeerDecimal(reader, "precio"),
            Activo = LeerBooleano(reader, "activo")
        };
    }
}
